rfcs/proposed/core_types/README.md

```diff
@@ -0,0 +1,285 @@
+real-world applications running on processors with more than two core types (e.g., on a system with performance (P),
+efficient (E), and low power efficient (LP E) cores):
+
+#### 1. **Flexibility and Resource Utilization**
+
+Many parallel workloads can execute efficiently on multiple core types. For example:
+- A parallel algorithm with good scalability works well on both P-cores and E-cores
+- Background processing can run on E-cores or LP E-cores depending on availability
+- Mixed workloads benefit from utilizing any available performance-class cores (P or E)
+
+Restricting to a single core type may leave available cores idle, reducing overall system throughput.
+
+#### 2. **Workload Classification Challenges**
+
+Applications often have workloads that don't fit neatly into a single core type category:
+- **Moderate priority tasks**: Not critical enough to demand P-cores exclusively, but shouldn't use LP E-cores
+- **Adaptive workloads**: Performance requirements that vary based on input size or system state
+- **Mixed computation phases**: Algorithms that alternate between compute-intensive and throughput-oriented phases
+
+#### 3. **Avoiding Inappropriate Core Selection**
+
+Without the ability to specify "P-cores OR E-cores (but not LP E-cores)", applications face a dilemma:
+- **No constraint**: Work might be scheduled on LP E-cores, causing significant performance degradation
+- **P-cores only**: Leaves E-cores idle, reducing parallelism
+- **E-cores only**: Misses opportunities to use faster P-cores when available
+
+### Current API Limitation
+
+The existing API only supports single core type constraints:
+
+```cpp
+auto core_types = tbb::info::core_types();
+// Assume: [0] = LP E-core, [1] = E-core, [2] = P-core
+
+tbb::task_arena arena(
+    tbb::task_arena::constraints{}.set_core_type(core_types[2])  // Only P-cores
+);
+```
+
+This forces applications to choose one of these suboptimal strategies:
+
+| Strategy | Pros | Cons |
+|----------|------|------|
+| **P-cores only** | Maximum single-threaded performance | Leaves E-cores idle; limited parallelism; higher power |
+| **E-cores only** | Good for parallel workloads | Doesn't utilize P-core performance; excludes LP E-cores |
+| **LP E-cores only** | Minimal power consumption | Severe performance impact for most workloads |
+| **No constraint** | Maximum flexibility | May schedule on inappropriate cores (e.g., LP E-cores for compute) |
+
+None of these options provide the desired behavior: **"Use P-cores or E-cores, but avoid LP E-cores"** or **"Use any
+efficiency cores (E-core or LP E-core)"**.
+
+## Proposal
+
+We propose extending the `task_arena::constraints` API to support specifying multiple acceptable core types, enabling
+applications to define flexible core type policies that adapt to workload requirements and available hardware
+resources.
+
+### New API
+
+Add the following methods to `tbb::task_arena::constraints`:
```

Comment:
Another possibility can be to leave `constraints` struct unchanged, but introduce, for example, a new constructor for the `task_arena` class that would accept vector/array of `constraints` instances, each bound to certain NUMA, core type and threads per core constraints. The task arena constructor would make a union of the masks resulting from each of `constraints` instance and use that union as its constraint.
At first glance, this design looks more flexible to me as it scales better, allowing users not only specifying more than one core type, but also more than one NUMA node. Essentially, users can specify multiple portions of the platform, whose united constraint is desired to be set for a single `task_arena` instance.

Reply:
Sounds like TCM. 😉 I will try to add this as an alternative.